### PR TITLE
Pricing contract

### PR DESCRIPTION
**Pull request by namnm1991**

### CODE REVIEW

File: reserve_sdk/contract.py [hunk 1]
```diff
@@ -1,7 +1,86 @@
...
+
+    Args:
+        rate: value of new sell/buy price
+        base: value of current sell/buy price at contract
+
+    Returns:
+        compact_data which include new base rate & compact value which is the
+        different between rate and base in bps unit.
+
+    TODO consider to raise BaseChangedException to indicates that base rates
```
[the remark this answers is not in the document]
Comment (namnm1991) [reply reserve_sdk/contract.py]: agree, no need to raise the exception in this function, also add base_changed property to CompactData is good idea

File: reserve_sdk/contract.py [hunk 2]
```diff
@@ -1,7 +1,86 @@
...
+    Args:
+        prices: price change in bps unit
+        token_indices: index of token in compact data on contract
+
+    Return:
+        buy: buy prices change in bps unit, encoded in hex
+        sell: sell prices change in bps unit, encoded in hex
+        indices: the index of block token in compact data on contract
+
+    TODO improve function documentation with example.
```
[the remark this answers is not in the document]
Comment (namnm1991) [reply reserve_sdk/contract.py]: I'll remove this TODO and add test case for this function, the test is a kind of example for useage

File: reserve_sdk/contract.py [hunk 3]
```diff
@@ -1,7 +1,86 @@
...
     RESERVE_CODE, CONVERSION_RATES_CODE, SANITY_RATES_CODE)
+from .utils import send_transaction, hexlify
+
+
+"""Show token position in the compact data."""
+TokenIndex = namedtuple('TokenIndex', ('array_idx', 'field_idx'))
+CompactData = namedtuple('CompactData', ('base', 'compact'))
+
+
+def get_compact_data(rate, base):
```
[the remark this answers is not in the document]
Comment (namnm1991) [reply reserve_sdk/contract.py]: sure, I will add more tests

File: reserve_sdk/contract.py [hunk 4]
```diff
@@ -1,7 +1,86 @@
...
+    if compact <= -128 or compact >= 127:  # not fit in a byte
+        return CompactData(rate, 0)
+    else:
+        # handle negative value to convert to byte
+        if compact < 0:
+            compact += 256
+        return CompactData(base, compact)
+
+
+def build_compact_price(prices, token_indices):
```
[the remark this answers is not in the document]
Comment (namnm1991) [reply reserve_sdk/contract.py]: sure

File: reserve_sdk/contract.py [hunk 5]
```diff
@@ -177,55 +302,273 @@ class ConversionRatesContract(BaseContract):
...
-    def set_sell_rate(self, token, rate):
-        """Place the selling rate for given token.
         Args:
             token: the token address
-            rate: new rate to set
+            buy: token buy price
+            sell: token sell price
+
+        Returns:
+            token: the token address # TODO consider to remove this output
```
[the remark this answers is not in the document]
Comment (namnm1991) [reply reserve_sdk/contract.py]: I will remove this.

File: reserve_sdk/contract.py [hunk 6]
```diff
@@ -177,55 +302,273 @@ class ConversionRatesContract(BaseContract):
...
+            1. Add token address to pricing contract.
+            2. Set token control info .
+            3. Enable token trade.
+        """
+        self.add_token(token)
+        self.set_token_control_info(
+            token, minimal_record_resolution,
+            max_per_block_imbalance, max_total_imbalance
+        )
+        self.enable_token_trade(token)
```
[the remark this answers is not in the document]
Comment (favadi) [reply reserve_sdk/contract.py]: I fear that prefetch all the token indices might be slow. Could you test?

File: reserve_sdk/contract.py [hunk 7]
```diff
@@ -100,65 +205,85 @@ def __init__(self, provider, account, address):
...
+
         Args:
             token: token address
             amount: amount of token to withdraw
             dest: destination address to receive the token
         """
-        raise NotImplementedError
+        tx = self.contract.functions.withdraw(
+            token, amount, dest).buildTransaction({
+                'gas': 3000000
```
[the remark this answers is not in the document]
Comment (namnm1991) [reply reserve_sdk/contract.py]: Yes, I used fixed gas in several place, will replace with estimate gas.

File: reserve_sdk/token.py [hunk 8]
```diff
@@ -0,0 +1,18 @@
+from .utils import call_contract
+
+
+class Token():
```
Comment (favadi) [inline reserve_sdk/token.py]: could you add a comment doc for this class? Also, should it be named ERC20 token?

**Pull request merged**